rust/noosphere-storage/src/exportable.rs

```diff
@@ -0,0 +1,176 @@
+use crate::{Storage, Store, SPHERE_DB_STORE_NAMES};
+use anyhow::Result;
+use async_trait::async_trait;
+use noosphere_common::ConditionalSync;
+use std::pin::Pin;
+use tokio_stream::{Stream, StreamExt};
+
+#[cfg(target_arch = "wasm32")]
+pub type IterableStream<'a> = Pin<Box<dyn Stream<Item = Result<(Vec<u8>, Option<Vec<u8>>)>> + 'a>>;
+#[cfg(not(target_arch = "wasm32"))]
+pub type IterableStream<'a> =
+    Pin<Box<dyn Stream<Item = Result<(Vec<u8>, Option<Vec<u8>>)>> + Send + 'a>>;
+
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait Iterable: Store {
+    fn get_all_entries<'a>(&'a self) -> IterableStream<'a>;
+}
+
+/// A blanket implementation for [Storage]s that can be
+/// imported via [Importable::import].
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait Exportable
+where
+    Self: Storage,
+    <Self as Storage>::KeyValueStore: Iterable,
+{
+    async fn get_all_store_names(&self) -> Result<Vec<String>> {
+        let mut names = vec![];
+        names.extend(SPHERE_DB_STORE_NAMES.iter().map(|name| String::from(*name)));
+        Ok(names)
+    }
+}
+
+impl<S> Exportable for S
+where
+    S: Storage,
+    S::KeyValueStore: Iterable,
+{
+}
+
+/// A blanket implementation for all [Storage]s to import
+/// an [Exportable] storage.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait Importable<'a, E>
+where
+    Self: Storage,
+    Self::KeyValueStore: Store,
```

Comment:
All `Store`s implement `KeyValueStore`, maybe `Store` should implement `as_block_store()`/`as_key_store()` rather than `Storage` distinguish different `Store` wrappers which aren't used internally

Reply:
This proposal breaks down with some storage layers e.g. `IpfsStorage` which rely on distinguishing block from KV stores